Source/SwiftLintBuiltInRules/Rules/Idiomatic/FileNameRule.swift

```diff
@@ -56,33 +56,98 @@ struct FileNameRule: OptInRule, SourceKitFreeRule {
 }
 
 private class TypeNameCollectingVisitor: SyntaxVisitor {
+    // All of a visited node's ancestor type names if that node is nested, starting with the furthest
+    // ancestor and ending with the direct parent
+    private var ancestorNames: [String] = []
+
+    // All of the type names found in the file
     private(set) var names: Set<String> = []
 
+    // If true, nested types are only allowed in the file name when used by their fully-qualified name
+    // (e.g. `My.Nested.Type` and not just `Type`)
+    private let requireFullyQualifiedNames: Bool
+
+    init(requireFullyQualifiedNames: Bool) {
+        self.requireFullyQualifiedNames = requireFullyQualifiedNames
+        super.init(viewMode: .sourceAccurate)
+    }
+
+    private func addVisitedNodeName(_ name: String) {
+        let fullyQualifiedName = (ancestorNames + [name]).joined(separator: ".")
+        names.insert(fullyQualifiedName)
+
+        if !requireFullyQualifiedNames {
+            names.insert(name)
+        }
+    }
+
+    override func visit(_ node: ClassDeclSyntax) -> SyntaxVisitorContinueKind {
+        ancestorNames.append(node.name.text)
```

Comment:
Move this line into `addVisitedNodeNames` and call it right here.

Reply:
Good point. I'm gonna make some changes to remove a lot of duplicate code and put them into common functions